src/systems/label/Label.cc

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include "Label.hh"
+#include <string>
+
+#include <ignition/plugin/Register.hh>
+#include "ignition/gazebo/EntityComponentManager.hh"
+#include "ignition/gazebo/components/SemanticLabel.hh"
+#include "ignition/gazebo/components/Link.hh"
+#include "ignition/gazebo/components/Visual.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+//////////////////////////////////////////////////
+Label::Label() : System()
+{
+}
+
+//////////////////////////////////////////////////
+Label::~Label() = default;
+
+//////////////////////////////////////////////////
+void Label::Configure(const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    gazebo::EntityComponentManager &_ecm,
+    gazebo::EventManager & /*_eventMgr*/)
+{
+  const std::string labelTag = "label";
+
+  if (!_sdf->HasElement(labelTag))
+  {
+    ignerr << "Failed to load Label system; label tag not found.\n";
+    return;
+  }
+
+  auto label = _sdf->Get<int>(labelTag);
+
+  if (label < 0 || label > 255)
+  {
+    ignerr << "Failed to configure Label system; value " << label
+      << " is not in [0-255] range.\n";
+    return;
+  }
+
+  const std::string parentName = _sdf->GetParent()->GetName();
+
+  // Set the component to the visual to set its user data, but
+  // if the plugin is inside the <model> tag, get its visual child
+  if (parentName == "visual")
+  {
+    _ecm.CreateComponent(_entity, components::SemanticLabel(label));
+  }
+  else if (parentName == "model")
+  {
+    // Get link childern of parent model
+    auto links = _ecm.ChildrenByComponents<components::Link>(
+      _entity, components::Link());
+
+    for (auto linkEntity : links)
+    {
+      // get visual child of parent link
+      auto visuals = _ecm.ChildrenByComponents<components::Visual>(
+        linkEntity, components::Visual());
+
+      // Create label component to all visual childern
+      for (auto visualEntity : visuals)
+      {
+        _ecm.CreateComponent(visualEntity,
```

Comment:
There is a problem with actors 

I am trying here to create a component of the <visual> of the model, so that i can set its user data in the RenderUtil.cc

Anyway, there is a problem with actor models only, as actors SDF has `<actor>` tag instead of `<model>` tag, so i cannot access its `<visual>` tag, so how can i do that ?

Reply:
> Anyway, there is a problem with actor models only, as actors SDF has <actor> tag instead of <model> tag, so i cannot access its <visual> tag, so how can i do that ?

I took a look, but I'm not sure either. Looking at the [actor SDF spec](http://sdformat.org/spec?ver=1.6&elem=actor), actors can have links, but this is optional. The `<skin>` element description says the following:
```
Skin file which defines a visual and the underlying skeleton which moves it.
```

So, perhaps if  a `<skin>` tag is nested in the `<actor>` tag, a visual is stored somewhere? If so, I'm not sure where the visual would be though. I tried the following code with the [`actor.sdf` example world](https://github.com/ignitionrobotics/ign-gazebo/blob/main/examples/worlds/actor.sdf), but did not see any printed link or visual information:
```c++
    ignerr << "checking actors..." << std::endl << std::endl;
    _ecm.Each<components::Actor>(
        [&](const Entity &_entity, const components::Actor *)
        {
          ignerr << "Actor " << _entity << std::endl;

          // see if an actor has links. If so, see if visuals are attached
          // to these links
          auto links = _ecm.ChildrenByComponents<components::Link>(
              _entity, components::Link());
          for (const auto &link : links)
          {
            ignerr << "\ton link " << link << std::endl;
            auto visuals = _ecm.ChildrenByComponents<components::Visual>(
                link, components::Visual());
            for (const auto &visual : visuals)
              ignerr << "\t\thas visual " << visual << std::endl;
          }

          // see if visuals are attached to actors directly
          auto visuals = _ecm.ChildrenByComponents<components::Visual>(
              _entity, components::Visual());
          for (const auto &visual : visuals)
            ignerr << "\thas visual " << visual << std::endl;
          return true;
        });
```

Maybe someone else can chime in to give more thoughts/ideas about this issue :grin:

Reply:
@adlarkin is right, the closest that an actor has to a visual in the SDF is the skin. But when it comes to entities, we're not creating separate entities for the skin and animation, we only have a single entity for the whole actor. So I think it makes sense to attach the component directly to the actor entity.

Reply:
> So I think it makes sense to attach the component directly to the actor entity.

If this approach were to be taken, is there still a way to get the visual of the actor in `ign-rendering`? The reason why we are attaching a component to something here is so that we can pass this information along to rendering, so that rendering knows what label is associated with a certain visual

Reply:
> is there still a way to get the visual of the actor in ign-rendering? 

I think you can update `SceneManager::NodeById` to also return actor nodes.

Reply:
I don't think that will work :disappointed: `SceneManager::NodeById` returns a `rendering::NodePtr`, but the actors are of type `rendering::MeshPtr`: https://github.com/ignitionrobotics/ign-gazebo/blob/f0e235b9c65f84128988648e9f3111ca7ed82c50/src/rendering/SceneManager.cc#L83

The problem is that a mesh is a geometry, and a geometry is an object... so, it doesn't look like the inheritance structure works to allow for this, since a mesh doesn't inherit from the node class at all

Reply:
Oh I just realized that actor visuals (which contain those meshes) are already tracked together with other visuals:

https://github.com/ignitionrobotics/ign-gazebo/blob/1e1d0bab44b516d8ae419f3f614630c5a20f82ab/src/rendering/SceneManager.cc#L990-L997

So it may already work

Reply:
Thanks for that conversation :D @chapulina @adlarkin